Library/Homebrew/cmd/upgrade.rb

```diff
@@ -104,6 +105,35 @@ def upgrade
         onoe "#{f}: #{e}"
       end
     end
+
+    unless Homebrew.failed?
+      #  Assess existing kegs and reinstall any broken.
+      broken_formulae = []
+
+      CacheStoreDatabase.use(:linkage) do |db|
+        kegs = Formula.installed.collect(&:opt_or_installed_prefix_keg).reject(&:nil?)
+
+        odebug "Assessing keg linkage status" if kegs.size > 1
+
+        kegs.each do |keg|
```

Comment:
Could consider slimming this down to only those formulae with `runtime_dependencies` on this formula (which indicates linkage already and is much faster than reading linkage on all kegs). Thoughts?

Reply:
```
Formula.installed
  .collect(&:opt_or_installed_prefix_keg)
  .keep_if {...}
```
Two options for checking here:
- `|k| k.runtime_dependencies.contains?(f)`;
- `|k| k.runtime_dependencies.map(&:name).include?(f.full_alias_name)`.

I went for the first one, but let me know which is correct!